components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -0,0 +1,453 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! The handling of `PrepareMerge` command.
+//!
+//! ## Propose (`Peer::propose_prepare_merge`)
+//!
+//! Checks for these requirements:
+//!
+//! - Validate the request. (`Peer::validate_prepare_merge_command`)
+//! - Log gap between source region leader and peers is not too large. This is
+//!   because these logs need to be embeded in the later `CommitMerge` command.
+//! - Logs that aren't fully committed (to all peers) does not contains
+//!   `CompactLog` or certain admin commands.
+//!
+//! Then, transfer all in-memory pessimistic locks to the target region as a
+//! Raft proposal. To guarantee the consistency of lock serialization, we might
+//! need to wait for some in-flight logs to be applied. During the wait, all
+//! incoming write proposals will be rejected. Read the comments of
+//! `MergeContext::prepare_fence` for more details.
+//!
+//! ## Apply (`Apply::apply_prepare_merge`)
+//!
+//! Increase region epoch and write the merge state.
+//!
+//! ## On Apply Result (`Peer::on_apply_res_prepare_merge`)
+//!
+//! Start the tick (`Peer::on_check_merge`) to periodically check the
+//! eligibility of merge.
+
+use engine_traits::{KvEngine, RaftEngine, RaftLogBatch, CF_LOCK};
+use kvproto::{
+    raft_cmdpb::{
+        AdminCmdType, AdminRequest, AdminResponse, CmdType, PrepareMergeRequest, PutRequest,
+        RaftCmdRequest, Request,
+    },
+    raft_serverpb::{MergeState, PeerState, RegionLocalState},
+};
+use parking_lot::RwLockUpgradableReadGuard;
+use protobuf::Message;
+use raft::{eraftpb::EntryType, GetEntriesContext, NO_LIMIT};
+use raftstore::{
+    coprocessor::RegionChangeReason,
+    store::{metrics::PEER_ADMIN_CMD_COUNTER, util, LocksStatus, ProposalContext, Transport},
+    Error, Result,
+};
+use slog::{debug, info};
+use tikv_util::{box_err, store::region_on_same_stores};
+
+use super::PrepareStatus;
+use crate::{
+    batch::StoreContext,
+    fsm::ApplyResReporter,
+    operation::AdminCmdResult,
+    raft::{Apply, Peer},
+    router::CmdResChannel,
+};
+
+#[derive(Debug)]
+pub struct PrepareMergeResult {
+    pub region_state: RegionLocalState,
+    pub state: MergeState,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn propose_prepare_merge<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        if self.storage().has_dirty_data() {
+            return Err(box_err!(
+                "{:?} source peer has dirty data, try again later",
+                self.logger.list()
+            ));
+        }
+        self.validate_prepare_merge_command(
+            store_ctx,
+            req.get_admin_request().get_prepare_merge(),
+        )?;
+        let (req, lock_size_limit) = self.pre_propose_prepare_merge(store_ctx, req)?;
```

Comment:
It's a waste to execute the check twice as it needs to parse logs.

Reply:
Will `min_matched` change during the wait?

Reply:
It can only be increased.